File: web/src/lib/events.ts
/**
 * The typed SSE event contract.
 *
 * This file mirrors backend/api/events.py. The web rules say change both
 * or neither: if you add an event type here, add it there and to the
 * spec in the same commit.
 */

export type StepStart = {
  type: "step_start";
  step: number;
  tool: string;
  args: Record<string, unknown>;
};

export type StepResult = {
  type: "step_result";
  step: number;
  summary: string;
};

export type Token = { type: "token"; text: string };

export type Citation = {
  type: "citation";
  path: string;
  start: number;
  end: number;
};

export type Done = {
  type: "done";
  steps: number;
  tokens: number;
  ms: number;
  cost_usd: number;
};

export type Progress = {
  type: "progress";
  stage: string;
  message: string;
  percent: number | null;
};

export type ErrorEvent = {
  type: "error";
  code: string;
  message: string;
  retryable: boolean;
};

export type Event =
  | StepStart
  | StepResult
  | Token
  | Citation
  | Done
  | Progress
  | ErrorEvent;

/**
 * Read an SSE body as typed events, yielding each as it arrives.
 *
 * EventSource cannot POST, and asking a question is a POST, so the
 * stream is read off fetch's ReadableStream instead.
 */
export async function* readEvents(
  response: Response,
  signal?: AbortSignal,
): AsyncGenerator<Event> {
  const body = response.body;
  if (!body) return;
  const reader = body.pipeThrough(new TextDecoderStream()).getReader();
  let buffer = "";

  try {
    while (true) {
      if (signal?.aborted) return;
      const { done, value } = await reader.read();
      if (done) break;
      buffer += value;

      // Frames are separated by a blank line; a partial frame stays in
      // the buffer until the rest of it arrives.
      let boundary = buffer.indexOf("\n\n");
      while (boundary !== -1) {
        const frame = buffer.slice(0, boundary);
        buffer = buffer.slice(boundary + 2);
        const line = frame.split("\n").find((l) => l.startsWith("data: "));
        if (line) yield JSON.parse(line.slice(6)) as Event;
        boundary = buffer.indexOf("\n\n");
      }
    }
  } finally {
    reader.cancel().catch(() => {});
  }
}

File: backend/api/events.py
"""The typed SSE event contract (docs/SPEC.md, "Streaming contract").

These models ARE the contract: the shapes here, the names in
web/lib/events.ts, and the spec block must stay identical. Adding an
event type means editing the spec in the same commit (rules/api.md).
"""

from __future__ import annotations

import json
from typing import Literal

from pydantic import BaseModel


class StepStart(BaseModel):
    """A tool call is about to run."""

    type: Literal["step_start"] = "step_start"
    step: int
    tool: str
    args: dict


class StepResult(BaseModel):
    """A tool call finished; `summary` is the one-line trace text."""

    type: Literal["step_result"] = "step_result"
    step: int
    summary: str


class Token(BaseModel):
    """A piece of the answer text."""

    type: Literal["token"] = "token"
    text: str


class Citation(BaseModel):
    """A validated `path:start-end` the answer relies on."""

    type: Literal["citation"] = "citation"
    path: str
    start: int
    end: int


class Done(BaseModel):
    """Terminal event for a successful run, with its accounting."""

    type: Literal["done"] = "done"
    steps: int
    tokens: int
    ms: int
    cost_usd: float = 0.0


class Progress(BaseModel):
    """Indexing progress: one event per pipeline stage boundary."""

    type: Literal["progress"] = "progress"
    stage: str
    message: str
    percent: int | None = None


class ErrorEvent(BaseModel):
    """A stream that dies must emit this before closing (rules/api.md)."""

    type: Literal["error"] = "error"
    code: str
    message: str
    retryable: bool


Event = StepStart | StepResult | Token | Citation | Done | Progress | ErrorEvent


def sse(event: BaseModel) -> str:
    """Serialize one event as an SSE `data:` frame."""
    return f"data: {json.dumps(event.model_dump())}\n\n"
